guides/keras_tuner/custom_tuner.py

```diff
@@ -161,53 +131,81 @@ def run_train_step(data):
+            print(f"Epoch: {epoch}")
 
-            self.on_epoch_begin(trial, model, epoch, logs={})
-            for batch, data in enumerate(train_ds):
-                self.on_batch_begin(trial, model, batch, logs={})
-                batch_loss = float(run_train_step(data))
-                self.on_batch_end(trial, model, batch, logs={"loss": batch_loss})
+            # Iterate the training data to run the training step.
+            for images, labels in train_ds:
+                run_train_step(images, labels)
 
-                if batch % 100 == 0:
-                    loss = epoch_loss_metric.result().numpy()
-                    print("Batch: {}, Average Loss: {}".format(batch, loss))
+            # Iterate the validation data to run the validation step.
+            for images, labels in validation_data:
+                run_val_step(images, labels)
 
-            epoch_loss = epoch_loss_metric.result().numpy()
-            self.on_epoch_end(trial, model, epoch, logs={"loss": epoch_loss})
+            # Calling the callbacks after epoch.
+            epoch_loss = float(epoch_loss_metric.result().numpy())
+            for callback in callbacks:
+                # The "my_metric" is the objective passed to the tuner.
+                callback.on_epoch_end(epoch, logs={"my_metric": epoch_loss})
             epoch_loss_metric.reset_states()
 
+            print(f"Epoch loss: {epoch_loss}")
+            best_epoch_loss = min(best_epoch_loss, epoch_loss)
 
-tuner = MyTuner(
-    oracle=kt.oracles.BayesianOptimization(
-        objective=kt.Objective("loss", "min"), max_trials=2
-    ),
-    hypermodel=build_model,
+        # Return the evaluation metric value.
+        return best_epoch_loss
+
+
+"""
+Now, we can initialize the tuner. Here we use `Objective("my_metric", "min")`
+as our metric to be minimized. It doesn't matter what name you use for the
+objective as long as it is consistent with the one you use as the key in the
+`logs` passed to 'on_epoch_end()' of the callbacks. The callbacks need to use
+this value in the `logs` to find the best epoch to checkpoint the model.
+
+"""
+tuner = kt.RandomSearch(
+    objective=kt.Objective("my_metric", "min"),
+    max_trials=2,
+    hypermodel=MyHyperModel(),
     directory="results",
-    project_name="mnist_custom_training",
+    project_name="custom_training",
+    overwrite=True,
 )
 
-(x_train, y_train), (x_test, y_test) = tf.keras.datasets.mnist.load_data()
 
-# Reshape the images to have the channel dimension.
-x_train = x_train.reshape(x_train.shape + (1,))[:1000]
-y_train = y_train.astype(np.int64)[:1000]
+"""
+We start the search by passing the arguments we defined in the signature of
+`MyHyperModel.fit()` to `tuner.search()`.
 
-mnist_train = tf.data.Dataset.from_tensor_slices((x_train, y_train))
+"""
+
+tuner.search(x=x_train, y=y_train, validation_data=(x_val, y_val))
 
-tuner.search(train_ds=mnist_train)
+"""
+Finally, we can retrieve the results.
```

Comment:
At the end of the guide, please add a couple of sentences with key takeaways from the guide and pointers to other resources.